### PR TITLE
gh-116023: Add `show_empty=False` to `ast.dump`

### PR DESCRIPTION
**Pull request by sobolevn**
My thoughts:
- This parameter should be optional and switched off by default, so we preserve the output if people rely on it: as debugging tool or somewhere in tests
- It should not touch attributes, because they already have a different flag

* Issue: gh-116023

----
📚 Documentation preview 📚: https://cpython-previews--116037.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/ast.rst [hunk 1]
```diff
@@ -2439,9 +2439,15 @@ and classes for traversing abstract syntax trees:
    indents that many spaces per level.  If *indent* is a string (such as ``"\t"``),
    that string is used to indent each level.
 
+   If *show_empty* is ``False``, then empty lists, fields that are ``None``,
+   and empty strings will be ommitted from the output for better readability.
```
Comment (AlexWaygood) [inline Doc/library/ast.rst]:
```suggestion
   and empty strings will be omitted from the output for better readability.
```

File: Lib/ast.py [hunk 2]
```diff
@@ -178,6 +186,7 @@ def _format(node, level=0):
         raise TypeError('expected AST, got %r' % node.__class__.__name__)
     if indent is not None and not isinstance(indent, str):
         indent = ' ' * indent
+    empty_values = {None, "''", "[]"}
```
Comment (JelleZijlstra) [inline Lib/ast.py]: I don't think we should omit empty strings, since we don't use them as defaults:

```
>>> n=ast.Name(ctx=ast.Load())
<stdin>-1:1: DeprecationWarning: Name.__init__ missing 1 required positional argument: 'id'. This will become an error in Python 3.15.
>>> n.id
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
    n.id
AttributeError: 'Name' object has no attribute 'id'
```

File: Lib/ast.py [hunk 3]
```diff
@@ -149,6 +155,8 @@ def _format(node, level=0):
                     keywords = True
                     continue
                 value, simple = _format(value, level)
+                if not show_empty and value in empty_values:
```
Comment (JelleZijlstra) [inline Lib/ast.py]: Why not check this before the `_format` call? That way we don't have to look at the string value.

File: Lib/ast.py [hunk 4]
```diff
@@ -114,7 +114,11 @@ def _convert(node):
     return _convert(node_or_string)
 
 
-def dump(node, annotate_fields=True, include_attributes=False, *, indent=None):
+def dump(
+    node, annotate_fields=True, include_attributes=False,
+    *,
+    indent=None, show_empty=True,
```
Comment (JelleZijlstra) [inline Lib/ast.py]: I was hoping we could jump to omitting them by default. `dump()` already changes regularly as new versions change the AST structure (e.g., 3.12 added `type_params` to classes and functions).

File: Lib/test/test_ast.py [hunk 5]
```diff
@@ -1334,6 +1334,51 @@ def test_dump_incomplete(self):
...
+        )
+
+        def check_text(code, expected, **kwargs):
+            nodes = ast.parse(code)
+            self.assertEqual(
+                ast.dump(nodes, show_empty=False, **kwargs),
+                expected,
+            )
+
+        check_text(
```
Comment (JelleZijlstra) [inline Lib/test/test_ast.py:1412]: Can you add a test involving None? I think we may need to add a special case so it shows `ast.Constant(value=None)` instead of just `ast.Constant()`:

```
>>> ast.Constant()
<stdin>-3:1: DeprecationWarning: Constant.__init__ missing 1 required positional argument: 'value'. This will become an error in Python 3.15.
  ast.Constant()
<ast.Constant object at 0x103880c90>
```

**Review by JelleZijlstra — COMMENTED** [review]
Thank you for picking this up!

File: Lib/ast.py (hunk 4 above)
Comment (sobolevn) [reply Lib/ast.py]: I am 50/50 on that :(

First of all, it really helps to explore the ast structure (I use `ast.dump` a lot in debugging https://github.com/wemake-services/wemake-python-styleguide).
Second, I think that it is ok to add new stuff (there's not other way), but it is not good to remove stuff by default that people can find useful.

But, I can easily be persuaded on switching the default :)

File: Doc/library/ast.rst [hunk 6]
```diff
@@ -2439,9 +2439,15 @@ and classes for traversing abstract syntax trees:
    indents that many spaces per level.  If *indent* is a string (such as ``"\t"``),
    that string is used to indent each level.
 
+   If *show_empty* is ``False``, then empty lists, fields that are ``None``,
+   and empty strings will be omitted from the output for better readability.
```
Comment (JelleZijlstra) [inline Doc/library/ast.rst]:
```suggestion
   If *show_empty* is ``False``, then empty lists and fields that are ``None``
   will be omitted from the output for better readability.
```

File: Lib/ast.py [hunk 7]
```diff
@@ -125,6 +129,8 @@ def dump(node, annotate_fields=True, include_attributes=False, *, indent=None):
     include_attributes can be set to true.  If indent is a non-negative
     integer or string, then the tree will be pretty-printed with that indent
     level. None (the default) selects the single line representation.
+    If show_empty is False, then empty lists, fields that are None,
+    and empty strings will be ommitted from the output for better readability.
```
Comment (JelleZijlstra) [inline Lib/ast.py]:
```suggestion
    If show_empty is False, then empty lists and fields that are None
    will be ommitted from the output for better readability.
```

File: Lib/ast.py [hunk 8]
```diff
@@ -148,6 +154,13 @@ def _format(node, level=0):
                 if value is None and getattr(cls, name, ...) is None:
                     keywords = True
                     continue
+                if (
+                    not show_empty
+                    and value in empty_values
```
Comment (JelleZijlstra) [inline Lib/ast.py]:
```suggestion
                    and (value is None or value == [])
```
I don't think the `empty_values` variable helps readability here

**Comment by sobolevn** [comment]
Updated! Thank you!

**Review by JelleZijlstra — APPROVED** [review]

**Comment by sobolevn** [comment]
Maybe @carljm or @hauntsaninja ? :)

**Review by hauntsaninja — COMMENTED** [review]
I think if we have the feature it should be on by default. Since it's sort of a convenience / conciseness feature, it's not as useful if it's off by default. Also seems fairly safe to keep on by default, since that's basically what 3.12's ast.dump will give you.

**Comment by carljm** [comment]
FWIW, I agree with @hauntsaninja and @JelleZijlstra. I think in the more common use cases, seeing empty values is noisy and not useful, and the arg can be passed if you want to see them. I don't think there are any guarantees around backwards-compatibility for the textual representation of an AST.

> that's basically what 3.12's ast.dump will give you.

That depends on whether you are talking about `ast.dump(ast.arguments())` (dumping a manually-constructed AST) or `ast.dump(ast.parse("def x(): pass"))` (dumping a parsed AST). For the latter case, `show_empty=False` is a new behavior that no previous Python version has ever had.

I still think it's the better behavior, though.

**Comment by sobolevn** [comment]
Ok, fair enough! I've changed the default from `show_empty=True` to `show_empty=False`.
I've also updated tests and docs. Please, take a look at my new changes :)

File: Lib/test/test_ast.py [hunk 9]
```diff
@@ -1227,21 +1227,20 @@ def test_dump(self):
...
         self.assertEqual(ast.dump(node),
             "Module(body=[Expr(value=Call(func=Name(id='spam', ctx=Load()), "
-            "args=[Name(id='eggs', ctx=Load()), Constant(value='and cheese')], "
-            "keywords=[]))], type_ignores=[])"
+            "args=[Name(id='eggs', ctx=Load()), Constant(value='and cheese')]))])"
         )
         self.assertEqual(ast.dump(node, annotate_fields=False),
             "Module([Expr(Call(Name('spam', Load()), [Name('eggs', Load()), "
-            "Constant('and cheese')], []))], [])"
+            "Constant('and cheese')]))])"
```
Comment (JelleZijlstra) [inline Lib/test/test_ast.py:1234]: We need to think more about the interaction between show_empty and annotate_fields=False.

The current behavior on this PR seems wrong:
```
>>> dump(arguments(args=[arg("x")]), annotate_fields=False)
"arguments([arg('x')])"
>>> dump(eval(_))
"arguments(posonlyargs=[arg(arg='x')])"
```

I think the right behavior should be that if annotate_fields=False, we show empty fields (even if show_empty=True) if there is a nonempty field following them.

Comment (sobolevn) [reply Lib/test/test_ast.py:1234]: Done, fixed!

File: Lib/test/test_ast.py [hunk 10]
```diff
@@ -1326,6 +1326,41 @@ def test_dump_incomplete(self):
...
+        self.assertEqual(ast.dump(node, annotate_fields=False),
+            "arguments([], [arg('x')])",
+        )
+        node = ast.arguments(posonlyargs=[ast.arg("x")])
+        self.assertEqual(ast.dump(node, annotate_fields=False),
+            "arguments([arg('x')])",
+        )
+        node = ast.arguments(posonlyargs=[ast.arg("x")], kwonlyargs=[ast.arg('y')])
+        self.assertEqual(ast.dump(node, annotate_fields=False),
+            "arguments([arg('x')], kwonlyargs=[arg('y')])",
```
Comment (JelleZijlstra) [inline Lib/test/test_ast.py:1340]: This is a bit weird (why is `kwonlyargs` still passed as a keyword arg) but it's consistent with how `annotate_fields` already works:

```
>>> ast.dump(ast.arguments(kwonlyargs=[ast.arg("x")]), annotate_fields=False)
"arguments(kwonlyargs=[arg('x')])"
```

So this seems fine.

**Review by JelleZijlstra — APPROVED** [review]

File: Lib/test/test_ast.py (hunk 10 above)
Comment (sobolevn) [reply Lib/test/test_ast.py:1340]: Because of these lines: https://github.com/python/cpython/blob/acf69e09c66f8473399fabab36b81f56496528a6/Lib/ast.py#L148-L150

```python
>>> import ast
>>> args = ast.arguments(kwonlyargs=[ast.arg('x')])
>>> args.vararg
>>> getattr(type(args), 'vararg', None)
```

**Comment by sobolevn** [comment]
Waiting for some time for @hauntsaninja and @carljm to provide feedback :)
Jelle, thanks a lot for the review and for finding `annotate_fields` bug!

File: Doc/library/ast.rst [hunk 11]
```diff
@@ -729,7 +722,10 @@ Comprehensions
 
    .. doctest::
 
-        >>> print(ast.dump(ast.parse('[x for x in numbers]', mode='eval'), indent=4))
+        >>> print(ast.dump(
+        ...     ast.parse('[x for x in numbers]', mode='eval'),
+        ...     indent=4, show_empty=True,
```
Comment (carljm) [inline Doc/library/ast.rst]: What is the logic behind which docs (doctests) here use `show_empty=True` and which don't?

I would expect none of them to use it, except for maybe one example demonstrating `show_empty=True` itself, where we document the argument.

File: Lib/ast.py [hunk 12]
```diff
@@ -125,6 +129,8 @@ def dump(node, annotate_fields=True, include_attributes=False, *, indent=None):
     include_attributes can be set to true.  If indent is a non-negative
     integer or string, then the tree will be pretty-printed with that indent
     level. None (the default) selects the single line representation.
+    If show_empty is False, then empty lists and fields that are None
+    will be ommitted from the output for better readability.
```
Comment (carljm) [inline Lib/ast.py]:
```suggestion
    will be omitted from the output for better readability.
```

File: Misc/NEWS.d/next/Library/2024-02-28-11-51-51.gh-issue-116023.CGYhFh.rst [hunk 13]
```diff
@@ -0,0 +1,2 @@
+Add ``show_empty=False`` parameter to :func:`ast.dump` to optionally enable
+showing empty fields in the AST.
```
Comment (carljm) [inline Misc/NEWS.d/next/Library/2024-02-28-11-51-51.gh-issue-116023.CGYhFh.rst]: This doesn't really clarify the behavior change.
```suggestion
Don't show empty fields (value `None` or `[]`) in :func:`ast.dump` by default. Add ``show_empty=False`` parameter to optionally show them.
```

**Review by carljm — COMMENTED** [review]
LGTM, modulo comments. Thanks!

File: Doc/library/ast.rst (hunk 11 above)
Comment (sobolevn) [reply Doc/library/ast.rst]: Hm, indeed. But, I think that we do the opposite. Since this is the place where every node is documented, we need to show all properties of all nodes. It would be strange to hide them here. And `show_empty=True` will also be shown.

Comment (sobolevn) [reply Doc/library/ast.rst]: Yeah, I've tried my suggestion and didn't like it. Now I agree with you, all should be minimized.

File: Doc/library/ast.rst [hunk 14]
```diff
@@ -2442,9 +2364,42 @@ and classes for traversing abstract syntax trees:
    indents that many spaces per level.  If *indent* is a string (such as ``"\t"``),
    that string is used to indent each level.
 
+   If *show_empty* is ``False``, then empty lists and fields that are ``None``
+   will be omitted from the output for better readability.
```
Comment (carljm) [inline Doc/library/ast.rst]: Sorry I didn't notice this the first time, but two nits here:

1) I think we can omit "for better readability" -- it's kind of obvious, and the job of the docs here is to explain what happens, not opine on why it's a good behavior.

2) This sentence reads a bit as if the default is `True`. It should either clarify that `False` is the default, or read the other way around (explain what happens if the flag is `True`).

```suggestion
   If *show_empty* is ``False`` (the default), empty lists and fields that are ``None``
   will be omitted from the output.
```

**Review by carljm — APPROVED** [review]
LGTM, one minor nit on the docs.

**Comment by sobolevn** [comment]
(Waiting for one more day)

**Comment by sobolevn** [comment]
Thanks everyone! 🎉

**Comment by sobolevn** [comment]
Maybe we should also mention this in `3.13.rst` in `ast` section?
